### ISSUE TITLE
Add ability to setup HTTPS on domains

### ISSUE THREAD

**Issue by danielsamuels** [onespacemedia/server-management#4]

**Comment by danielsamuels** [comment]
Get certs from Let's Encrypt:

```
wget https://dl.eff.org/certbot-auto
chmod a+x certbot-auto
sudo ufw allow "Nginx HTTPS"
./certbot-auto --nginx (spacebar on the ones you don't want)
```

Cronjob:

```
56 3 * * * /home/deploy/certbot-auto renew --no-self-upgrade --quiet
```

Create DH Params:

```
cd /etc/ssl/
sudo openssl dhparam -out dhparams.pem 2048
```

New nginx config (replace the bit in the middle):

```
    ssl_session_timeout 1440m; # managed by Certbot

    # Diffie-Hellman parameter for DHE ciphersuites, recommended 2048 bits
    ssl_dhparam /etc/ssl/dhparams.pem;

    # modern configuration. tweak to your needs.
    ssl_protocols TLSv1.2;
    ssl_ciphers 'ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256';
    ssl_prefer_server_ciphers on;

    # HSTS (ngx_http_headers_module is required) (15768000 seconds = 6 months)
    add_header Strict-Transport-Security max-age=15768000;

    resolver 8.8.8.8;

    if ($scheme != "https") {
        return 301 https://$host$request_uri;
    } # managed by Certbot

```

**Comment by danielsamuels** [comment]
https://observatory.mozilla.org/analyze.html?host=www.touchstoneinnovations.com

**Comment by lewiscollard** [comment]
> X-Frame-Options (XFO) header not implemented

Interesting. Django has this in its default `startproject` - Is there any reason we should not enable this by default in the project template?

**Comment by danielsamuels** [comment]
No reason at all, we're just using a very old base settings file. We should probably take a look at the latest default and pull some things over.

**Comment by dan-gamble** [comment]
Might be worth doing that when `1.11` comes out and we update the cms

**Comment by danielsamuels** [comment]
Add `./manage.py ssl` command to automatically setup SSL.

**Comment by danielsamuels** [comment]
So it turns out that the version of OpenSSL that comes with 14.04 is too low to support ALPN HTTP/2.  I need to figure out if we can get it working on 14.04 (via compiling OpenSSL & Nginx, which would remove automatic updates), or if I should get the deployment script working with 16.04 LTS. Moving to 16.04 would probably be the best option.

I do have Let's Encrypt working though, so it might be worth splitting that out to get it in earlier.

Some useful links:

https://ma.ttias.be/day-google-chrome-disables-http2-nearly-everyone-may-31st-2016/
https://www.miguelvallejo.com/updating-to-openssl-1-0-2g-on-ubuntu-server-12-04-14-04-lts-to-stop-cve-2016-0800-drown-attack/
http://stackoverflow.com/questions/1210883/determine-supported-http-version-by-the-web-server

**Comment by lewiscollard** [comment]
Splitting out the LE stuff might be a good idea, yes.

On the other hand, if it turns out to be infeasible to port to 16.04 right now, OpenSSL 1.0.2 is [available](https://packages.debian.org/jessie-backports/openssl) in `jessie-backports`, which would kick the can down the road for a few years if we did the relatively-easy fixes in #44.

**Comment by danielsamuels** [comment]
Fixed on the [http2-lets-encrypt](https://github.com/onespacemedia/server-management/tree/http2-lets-encrypt) branch.
